### ISSUE TITLE
Manually add URL's to a container

### ISSUE THREAD

**Issue by TriMoon** [mozilla/multi-account-containers#1670]
- Multi-Account Containers Version: 6.2.3
- Operating System + Version: Kubuntu 19.10
- Firefox Version: Nightly `Mozilla/5.0 (X11; Linux x86_64; rv:75.0) Gecko/20100101 Firefox/75.0 ID:20200229214828`
<details>
<summary>Other installed Add-ons + Version + Enabled/Disabled-Status (Click to expand/hide)</summary>

About Sync 0.0.51
Amazon.com 1.1
Authenticator 6.0.1
Awesome RSS 1.3.5
Bing 1.1
CanvasBlocker 1.1 [DISABLED]
Certainly Something (Certificate Viewer) 1.2.3
Compat Report 0.1.0
Containers theme 1.0.3 [DISABLED]
Context Plus 0.6.0
Debian queries 2.3
Decentraleyes 2.0.13
Disconnect 5.19.3 [DISABLED]
DuckDuckGo 1.0
DuckDuckGo Privacy Essentials 2019.12.12 [DISABLED]
Emoji Cheatsheet for GitHub, Basecamp etc. 2.0.1
Enhancer for YouTube™ 2.0.99
Enterprise Policy Generator 5.1.0 [DISABLED]
Feedbro 4.0.3
Firefox DevTools ADB Extension 0.0.5
Firefox Lightbeam 2.1.0
Firefox Multi-Account Containers 6.2.3
FoxyProxy Standard 7.4.3
Google 1.0
Greasemonkey 4.9
HTTP/2 Indicator 3.0
IPFS Companion 2.10.0
Laboratory 3.0.7 [DISABLED]
MathML Font Settings 0.3.2
Native MathML 1.9.4
Nightly Tester Tools 4.0
Notes by Firefox 4.3.5 [DISABLED]
Offline QR Code Generator 1.6
PassFF 1.9.1 [DISABLED]
Plasma Integration 1.7
ProtonDB for Steam 1.6.0
React Developer Tools 4.4.0 [DISABLED]
Session Sync 3.1.12
Snooze Tabs 1.1.1
Tab Session Manager 5.2.0 [DISABLED]
To Google Translate 4.0.6
Tree Style Tab 3.3.6
True Sight 1.7.1
Twitter 1.0
Vue.js devtools 5.3.3
Wikipedia (en) 1.0
YouTube Container 1.3.0
eBay 1.0
uMatrix 1.4.0

</details>

### Actual behavior
No way to add redirecting URL's.
- **If you DON'T FORGET TO DO THIS EVERY TIME:**
  - Only way at moment is to use right-click and choose a container.
  - Else you are screwed...

### Expected behavior
To automatically open known redirect URI's we need a EASY way to add them to a container.
Sometimes you want to catch redirect links, to force them to always open in a certain container **before** the browser navigates to the destination of the link.
For this purpose we **NEED** a way to manually add URL's to a container of choice !

### Steps to reproduce
1. Try to add a rule using the UI for links like `https://go.microsoft.com/fwlink/?LinkID=xxxxx&wt.mc_id=fuckmedead`
2. Stuck because there is no way.
3. Write bug report on github, like i do now :wink: 

### Notes
Related issues: #719, #837, #839, #1057, #1075, #1180, #1335, #1317 and maybe many more im tired search now...

**Comment by fprochazka** [comment]
Yes please, this would make the extension amazing (now it's only great).

I want to be able to add amazon iam-account login links to containers, but they are redirects, which means there is no moment to intercept it and add as "always open" in a selected container.

This would allow me to:

**setup:**
* create new AWS IAM user, store the login link (which has account ID in its url) and login credentials to LastPass
* create a new Firefox Container for that AWS account, add the login link as always open in this container

**usage:**
* open LastPass window, search for the AWS account I want to log into, click the item
* LastPass opens the login link
* Firefox Containers intercepts the link and opens it appropriate container (to prevent login conflicts)
* LastPass fills in login credentials
* I submit the login form and I'm in the correct AWS account

**Comment by sherry13131** [comment]
Hi @kendallcorner, can you assign this issue to me?

**Comment by AdamJel** [comment]
Further to this:
The settings of 'limit to designated sites' in NOT currently synced. The site list is, but it is needed to tick the box again on new instance of Firefox. It would be very useful to synchronize this as well.
And I support the idea of having option to import/export/ settings = managing it by file, which may then be on a repo and thus having the ability to declaratively set the state. I already have quite a number of containers, so having the option to manage it by script (json) and having it back on gitxxx would be nice.

**Comment by B00ze64** [comment]
I believe this would be better served by being added to Firefox, on the Options page, rather than in the tiny extension pop-up menus. This way we can have a large text box where that can edit and copy/paste into. This would also make moving container assignment to new profiles a matter of copy/pasting.

**Comment by AdamJel** [comment]
> I believe this would be better served by being added to Firefox, on the Options page, rather than in the tiny extension pop-up menus. This way we can have a large text box that we can edit and copy/paste into. This would also make moving container assignment to new profiles a matter of copy/pasting.

Imagine you have 100+ containers. Then the `matter of copy/pasting` is not a pleasant option. On the other hand, one text file, beiing yaml, json or whatever, is easy enough.

And these two options dont exclude one another. It may be, that the window for copy pasting is just a gui for that text file.

For me, it would be extreamly pleasant to have just one descriptive text file and option to point extension to that file to load and set all containers, site lists included.

**Comment by B00ze64** [comment]
I agree with you. Never occurred to me someone would have that many containers lol, but of course **Export/Import** would be awesome. Right now settings are stored in IndexedDB. You can move that around (storage folder,) but because of ID's etc, it will probably not work in another profile. I still would appreciate a nice textbox, preferably in the options, where I can quickly add and remove stuff from within the browser, by just typing domains, and copy/paste stuff between containers etc. But if Import/Export uses yaml, json or whatnot that will do fine, we can edit that easily enough!

**Comment by jeremymeyers** [comment]
[Simple Tab Groups](Simple Tab Groups) handles this pretty well, and you can use a regex string too.

**Comment by xdhmoore** [comment]
@jeremymeyers I think your link got wrong-linked...

**Comment by jeremymeyers** [comment]
@xdhmoore my bad. fixed now.

**Comment by hmlendea** [comment]
I am struggling with this because of redirects... It would be awesome if we could do it, as well as allowing wildards such as `*.aliexpress.com`

**Comment by grahamperrin** [comment]
> #839

@TriMoon we have triplication of issues. Please, can you close this one? Leaving the automated cross-references in the earlier issues. Thank you.

**Comment by Korb** [comment]
I have mistakenly posted a similar proposal to [bugzilla.mozilla.org](https://bugzilla.mozilla.org/show_bug.cgi?id=1713357). The full text can be read at the link I gave. In general, I support the proposal to add the ability to add dozens and hundreds of domains with all subdomains to containers in the same way as it is done with custom filters in ad blockers.

**Comment by comrademmirg** [comment]
I'm surprise to find this isn't a feature, could have a simple check box "always open this site in this container". Seriously, who on earth would want to manually select containers every time they open a new tab?

**Comment by mc0e** [comment]
Why has this been waiting so long?  IT doesn't sound like there's disagreement with the proposal, and @sherry13131 offered to work on it.

**Comment by TriMoon** [comment]
> > #839
> 
> @TriMoon we have triplication of issues. Please, can you close this one? Leaving the automated cross-references in the earlier issues. Thank you.

There is a reason why issues stay open, even if duplicates exist.
(In which case it shows how wanted it is)

So I politely decline your request 😉

**Comment by kurahaupo** [comment]
I note that issue #1833 is currently pinned, so it's *much* more likely to receive attention than any of these related bugs.

**Comment by kurahaupo** [comment]
I note that @sherry13131 did work on it and appears to have, if not a complete solution, at least a step in the right direction. However the PR has gone stale for 16 months and so I've re-submitted it as PR #2114.

**Comment by sshquack** [comment]
Please provide this functionality to manually add a URL to a container. Here is my config:
- Firefox 95.0 (64-bit)
- MAC 8.0.3
- macOS monterey 12.0.1 on M1

**Comment by MadManMoon** [comment]
## Additional Vote For Functionality

I also would absolutely like this functionality.

FF 99 Nightly
Win 10, Linux (i know) - Latest

I would also suggest the ability for other add-ons to address this one simply via API. So that programs such as firewalls, or script blockers like noscript, can assist in management of sites, and provide even deeper sandboxing of scripts and more per container.

<hr/>

## The IndexedDB Database

Where is this IndexedDB file, and can I easily edit it in DB Browser (SQLite) or something? (_as a scrappy workaround_) I've had a look around (including the XPI) and can't find anything obvious.

Also, I have what I would've thought as more containers than most. Were 'ABE' still a thing in noscript (it will be again), I'd wholly like to have that and containers fully syncopatio.

> I agree with you. Never occurred to me someone would have that many containers lol, but of course **Export/Import** would be awesome. Right now settings are stored in IndexedDB.

**Comment by mc0e** [comment]
> Never occurred to me someone would have that many containers

The only reason I don't have hundreds of site specific containers is because of the awkwardness of the interface for making them work. This issue is a prime example.

**Comment by Genda1ph** [comment]
It's baffling how this is not implemented yet...

**Comment by Shadowfire-Elite** [comment]
@kurahaupo
Hi, has there been any update on this issue?

**Comment by kurahaupo** [comment]
@Shadowfire-Elite I've in the process of rebasing my PR #2114 yet again, as it was 300+ commits out of date.

It's pending approval, but so far nobody has seen fit even to comment, much less explain the lack of approval.

I suspect it might require some regression tests to be added, but I have no idea how to do that.  If anyone has any ideas on how to do that, that would be much appreciated.

**Comment by mrgum** [comment]
Can sites be generalized to url patterns/prefixes?

Use case:
AWS Console access via SSO

creates links to the management console in all accounts and permission sets i have access to in that SSO/Org
https://myorg.awsapps.com/start/#/saml/custom/1234567890

I'd like to be able to assign each account to different container rather than choosing it manually every time

**Comment by rymancl** [comment]
> Can sites be generalized to url patterns/prefixes?
> 
> Use case: AWS Console access via SSO
> 
> creates links to the management console in all accounts and permission sets i have access to in that SSO/Org https://myorg.awsapps.com/start/#/saml/custom/1234567890
> 
> I'd like to be able to assign each account to different container rather than choosing it manually every time

I had to use [Containerise](https://github.com/kintesh/containerise) in conjunction with Firefox MAC to achieve this. 
Azure AD example:
```
account.activedirectory.windowsazure.com/applications/signin/<prod-application-id> , AWS  Prod
account.activedirectory.windowsazure.com/applications/signin/<nonprod-application-id> , AWS  Nonprod
etc
etc
```

**Comment by rakheshster** [comment]
I use the [Open external links in a container](https://addons.mozilla.org/en-GB/firefox/addon/open-url-in-container/) extension for this. You need to specify the Url in this format: `ext+container:name=MyContainer&url='http://whatever'`.

So, for example, I have this as the bookmarked Url for Microsoft 365 admin center: `ext+container:name=Work+Admin&url=https://admin.microsoft.com/AdminPortal/Home#/homepage` Clicking on the bookmark opens this in the "Work Admin" container.

**Comment by ReticentAdjutant** [comment]
What's keeping this feature from being implemented? It's in the top three most commented issues and the related issue (#473), which is also in the top three, also hasn't been implemented. Having to use the two or three other container addons has been a huge waste of time as they still don't fix what problems that are integral to the purpose of containers. It's been almost two years -- are containers obsolete? Is that why it hasn't happened and I just missed it?

**Comment by xanoni** [comment]
I just came here to open exactly this feature request.

It's a major restriction that I can't manually edit the site list. For sites that redirect between subdomains frequently (think Google), this makes it cumbersome (sometimes impossible) to use the "Limit to designated sites" setting.

**Comment by xanoni** [comment]
> https://addons.mozilla.org/en-US/firefox/addon/simple-tab-groups/ handles this pretty well, and you can use a regex string too.
> 
> [EDIT to fix url]

This looked promising on first sight, has someone else tried this and can confirm it works well with this Multi Account Containers extension?

**Comment by achernyakevich-sc** [comment]
@xanoni I use Simple Tab Groups (STG) with Multi-Account Container (MAC) for years.

STG is great tool for tabs management, grouping, sorting as it is designed for it. As well STG has great integration with MAC (like sticky containers or opening container in special tab groups, etc.) This pair of add-ons make work in Firefox well organized.

**Comment by influential-eliot** [comment]
Hiya, @achernyakevich-sc, I've noted that STG has pretty strong issues relating to MAC which I've raised here:
https://github.com/Drive4ik/simple-tab-groups/issues/990
Essentially when you reopen STGs (sometimes during the same session) it'll often just flat out forget the MAC the tabs were in. This basically breaks MAC in a way, however when I raised this in the mozilla ticket below, they flagged it wrongly, I believe. This is because I believe that the issue highlights an exploit in MAC.
https://github.com/mozilla/multi-account-containers/issues/2453

I appreciate that others may not have either reported this (or maybe even experienced or noticed it), but it doesn't mitigate the fact that it happened.

Either way, whilst I desparately wanted STG to work (or just bring back Mozilla's tab groups!) I couldn't use it for work stuff, and I *certainly* couldn't use it for home stuff. Privacy wise it's a hard stop and a blocker.

> @xanoni I use Simple Tab Groups (STG) with Multi-Account Container (MAC) for years.
> 
> STG is great tool for tabs management, grouping, sorting as it is designed for it. As well STG has great integration with MAC (like sticky containers or opening container in special tab groups, etc.) This pair of add-ons make work in Firefox well organized.

**Comment by xanoni** [comment]
BTW, have we found a simple workaround yet that would allow us to manually add domains? E.g., by editing some JSON or SQLITE file outside the browser? Haven't had time to start grepping, yet, so wanted to check if someone else has already done so.

**Comment by achernyakevich-sc** [comment]
> BTW, have we found a simple workaround yet that would allow us to manually add domains? E.g., by editing some JSON or SQLITE file outside the browser? Haven't had time to start grepping, yet, so wanted to check if someone else has already done so.

@xanoni There is [Containerise](https://addons.mozilla.org/en-US/firefox/addon/containerise/) add-on that probably will resolve your issue even simpler way. :)

**Comment by xanoni** [comment]
> @xanoni There is [Containerise](https://addons.mozilla.org/en-US/firefox/addon/containerise/) add-on that probably will resolve your issue even simpler way. :)

Thank you, but I'm trying to keep my number of add-ons to a minimum (given security / complexity / etc.).

**Comment by bohdan-shulha** [comment]
Hi team,

Any updates on this?

Last year I started replacing Google stuff with some non-spying (/less-spying) alternatives. This year I want to focus on opensource or, at least, companies that contribute back to the community (Bitwarden was the very first in this journey).

I find Containers extremely useful (I have used to Chrome profiles and I like containers more), but the fact that I can't assign the redirect link to some container, makes them unsusable by a large factor.

I have seen a multiple PRs regarding this feature. One of them is still open - https://github.com/mozilla/multi-account-containers/pull/2114.

Why this issue is not fixed for years? This is not the case when community doesn't want to contribute - this is, for sure, a very important feature that have got attention from lots of community members and developers. It is not going to break UX for a regular (non-engineering) user, but it will give devoted users a real tool.

:(

**Comment by tsbertalan** [comment]
I am unable to open e.g. Google Photos without disabling "limit to designated sites" for my Google container. It insists on taking me to through some redirection sequence every time, even if the container is logged in, and I enter some direct link into the app. (This of course is some idiotic design on Google's part--you can't directly type photos.google.com into the address bar and be in the app without first seeing some asinine marketing/landing page and having to click a button.)

**Comment by xanoni** [comment]
> I am unable to open e.g. Google Photos without disabling "limit to designated sites" for my Google container. It insists on taking me to through some redirection sequence every time, even if the container is logged in, and I enter some direct link into the app. (This of course is some idiotic design on Google's part--you can't directly type photos.google.com into the address bar and be in the app without first seeing some asinine marketing/landing page and having to click a button.)

Yeah I think this is exactly the issue that others have been describing in various tickets over the past few years. Below is another workaround that I haven't seen mentioned, yet.

**Steps:**
1. Use your favorite search engine and look for "full list of Google subdomains" (should yield something like [this here](https://gist.github.com/abuvanth/b9fcbaf7c77c2954f96c6e556138ffe8))
2. Choose a list of relevant subdomains, including those involved in the redirection process (if you're fast you can see them in the URL bar, otherwise use the inspector to observe what's going on)
3. For each of those subdomains, navigate your browser to an invalid resource (e.g., for the `analytics` subdomain you could go to `https://analytics.google.com/foobar`)
4. This should give you a `404` page instead of a redirection ... now you can add this domain to your container URL list and repeat the process with the other subdomains

Painful, but only has to be done once (until Google changes something ;-).

**Comment by xanoni** [comment]
Referring to my previous comment, I realized that it's not always possible to trigger a `404` response. Some servers will just always redirect to a different subdomain, even if you navigate to an invalid resource.

In that case, you can just disable redirects in your browser temporarily. In Firefox, there's a setting called `network.http.redirection-limit` in `about:config` for that (see [here](https://kb.mozillazine.org/Network.http.redirection-limit)).

Just set it to `0` and you should be able to add the subdomain via the add-on menu. I would hope that a similar setting exists for Chrome-based browsers, but I haven't checked (should be a quick web search... good luck).

**Comment by mc0e** [comment]
While it would be useful to be able to manually add domains, it would still remain clumsy at best to identify the domains involved in redirects, navigate to the right containers and paste them in.

What would be really nice would be to be able to display a list of recent urls, (hopefully including the whole redirect chain that have been directed outside the container because they didn't match the domain list for the container.  I'd then want a clickable interface to add the domains that I'd prefer not to jump outside the originating container, and they'd get added to that container accordingly.

**Comment by tsbertalan** [comment]
@xanoni That 404 method sounds like just the workaround that will do it, thanks. I figured identify the particular domains would not be too difficult, especially since people will have been complaining about this particular problem already, and you can just pull the domains from their posts. 

@mc0e You're right that this would benefit from a better interface for getting the redirected domains. But, I expect that some sort of analysis paralysis is what has prevented this feature from being implemented, because it wasn't ever going to be implemented in the perfect way.

**Comment by influential-eliot** [comment]
Also, adding the domains to a container that is locked to the domains means that those domains will also ONLY open in that container ...

**Comment by philsnow** [comment]
Another workaround (because I found some sites where the 404 workaround doesn't apply) is to just go to the extension's inspector window and get/set the right records in the extension's localstorage.  Write-up here: https://gist.github.com/philsnow/322bea2867d401171ada1c2eb7a38c89

**Comment by tsbertalan** [comment]
Now, a couple months later--from @philsnow  's link, I used this snippet:

```javascript
await ( async (prev, next) => {
  const prevURL = "siteContainerMap@@_" + prev;
  const nextURL = "siteContainerMap@@_" + next;
  await browser.storage.local.set({
    [nextURL]: (await browser.storage.local.get(prevURL))[prevURL]
  });
})("www.example.com", "SOMEWHERE.example.com") 
```

replacing `www.example.com` with `www.google.com` and also replacing `SOMEWHERE.example.com` with

1. `login.youtube.com`, and
2. `accounts.youtube.com`

I was able to defeat the login loop on YouTube. I'm not sure whether `login` or `accounts` did it; I was just guessing. I already had `accounts.google.com` added using the `network.http.redirection-limit=0` trick, which was necessary but not sufficient.

**Comment by Zuccace** [comment]
> In that case, you can just disable redirects in your browser temporarily. In Firefox, there's a setting called `network.http.redirection-limit` in `about:config` for that (see [here](https://kb.mozillazine.org/Network.http.redirection-limit)).
> 
> Just set it to `0` and you should be able to add the subdomain via the add-on menu.

This seems to be the simplest route. Although I wouldn't call it simple by any means. But it works at least.

**Comment by influential-eliot** [comment]
Yeah, if were to allegorically speak in ITIL terms and say this whole thing was an incident (_I know it isn't_) ... then that is not a fix and it is *barely* a work around. 

Plus, it's not something that general FireFox users will consider, in that it woud not even factor into their UX. 

Sorry to just post negativity, it is not my usual thing. 😞 

@Zuccace [said](https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1879699951):
> This seems to be the simplest route. Although I wouldn't call it simple by any means. But it works at least.
> 
> @xanoni [said](https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1416930533):
> > In that case, you can just disable redirects in your browser temporarily. In Firefox, there's a setting called `network.http.redirection-limit` in `about:config` for that (see [here](https://kb.mozillazine.org/Network.http.redirection-limit)).
> > 
> > Just set it to `0` and you should be able to add the subdomain via the add-on menu. _(EDIT: deleted one sentence that didn't help in this context here, lol.)_

**Comment by Zuccace** [comment]
Manually adding hosts would surely help, but as I just added yet another site to its own container, I encountered another problem: even if I gradually increased the redirect limit I seemed to miss one domain. I eventually brought up the inspector and from there opened a site in a new tab (with 0 redirect on). From there it was easy to add that domain to the correct container.
So there seem to be some javascript redirects on some sites that don't adhere to the (temporarily set) no-redirects setting.

The solution would be to have something to track which domains a tab has connected, then select and add sites from that list to the container.

**Comment by influential-eliot** [comment]
Agreed 100% !
@Zuccace [said](https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1892534530):
> Containers leaking data is a security issue.

... ... although I could've sworn I saw Firefox handle the '_login_' part of all of this automatically the other day.

Basically, I was in one container, it opened the login in another, then closed that and refreshed the original to finish the login process.

So it's possible that Moz are in the midst of handling the '_login flow over a different domain_' thing which is a part of this multi-faceted request.

**Comment by E14** [comment]
> Containers leaking data *is* a security issue.

Agreed, but containers and MAC in particular seem to be going the tab groups route from the core team support I'm seeing here.

> Basically, I was in one container, it opened the login in another, then closed that and refreshed the original to finish the login process.

Maybe you were previously logged-in in the origin container and just authorized the app? Logins need to set cookies and those definitely should not be shared across containers, that's the whole point

**Comment by enikkari** [comment]
Easy workaround to manually add a redirecting url: 

1. disconnect internet / wifi
2. put in url bar the url & hit enter
3. site is of course not found or redirected, add to container
4. Reconnect internet, done :D 

I used this to add manually my different AWS accounts to the multi account containers. Or any other link where the url redirects
https://<account-name>.signin.aws.amazon.com/console

**Comment by influential-eliot** [comment]
Very helpful, but this is obviously not a reason to close this.

@enikkari [said](https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1919204152):
> Easy workaround to manually add a redirecting url:
> 
>     disconnect internet / wifi
>     put in url bar the url & hit enter
>     site is of course not found or redirected, add to container
>     Reconnect internet, done :D
> 
> I used this to add manually my different AWS accounts to the multi account containers. Or any other link where the url redirects
> https://.signin.aws.amazon.com/console

**Comment by kurahaupo** [comment]
@enikkari It's a good option for those who can do that, but "shutting off the internet connection" is unreasonable where it would disrupt persistent connections such as ssh or irc. A more targetted approach that only blocks *new* TCP connections would be preferable. One way to do this on Linux is using `iptables`. To start blocking use:

    sudo iptables -I OUTPUT -p tcp --syn -j REJECT --reject-with icmp-admin-prohibited

To unblock change `-I` to `-D`:

    sudo iptables -D OUTPUT -p tcp --syn -j REJECT --reject-with icmp-admin-prohibited

As an alternative , using `-j DROP` will work, but it's much slower as Firefox will have to wait for each connection attempt to time out.

**Comment by jhard** [comment]
Also affected by this issue, I might have a workaround that's easier. 
When you confirm adding a site to a specific container's config, that confirmation page's URL will be similar to `moz-extension://c1e32a09-c7fd-4fcb-94d2-fa7ff73ecf64/confirm-page.html?url=https%3A%2F%2Fexample.com&cookieStoreId=firefox-container-123`

You can replace the URL (and cookieStoreId) with whatever you want to include, and then confirm that, and it will be added correctly.

**Comment by influential-eliot** [comment]
Interesting, @jhard, thanks!

> Also affected by this issue, I might have a workaround that's easier. When you confirm adding a site to a specific container's config, that confirmation page's URL will be similar to `moz-extension://c1e32a09-c7fd-4fcb-94d2-fa7ff73ecf64/confirm-page.html?url=https%3A%2F%2Fexample.com&cookieStoreId=firefox-container-123`
> 
> You can replace the URL (and cookieStoreId) with whatever you want to include, and then confirm that, and it will be added correctly.

So, we'd just then need to know the container ID, one would assume. Plus to be cognisant of domain/subdomain stuff, of course.

**Comment by kurahaupo** [comment]
That almost sounds like I could hijack the "choose search engine" facility, since that is mostly just "paste the search term into the middle of this custom URL".

**Comment by Holbormon** [comment]
> Also affected by this issue, I might have a workaround that's easier. When you confirm adding a site to a specific container's config, that confirmation page's URL will be similar to `moz-extension://c1e32a09-c7fd-4fcb-94d2-fa7ff73ecf64/confirm-page.html?url=https%3A%2F%2Fexample.com&cookieStoreId=firefox-container-123`
> 
> You can replace the URL (and cookieStoreId) with whatever you want to include, and then confirm that, and it will be added correctly.

Interesting workaround. Unfortunately, it doesn't work in the specific case of two URLS with the same domain. Here is an example: 

I want to use the custom URL `https://outlook.office365.com/mail/?login_hint=example1@example.com` to assign Outlook's login page with a specific e-mail account (here `example1@example.com`) to a specific container. Since I have two containers and two e-mail addresses, I tried the above, but both URLs still open in one container: the last one I try the workaround on always prevails.

**Comment by kurahaupo** [comment]
@Holbormon that is not a shortcoming in this plug-in; it's by design in the core of Firefox.

The core container mechanism in Firefox only consults the domain (aka "site"), and so when you ask any plug-in (this one included) to add a URL to a container, it necessarily has everything but the domain stripped from it.

**Comment by repsejnworb** [comment]
Why is this not in yet?

This plugin loses soo much value without this, with all the redirect shenanigans on the internet..
Manual adding and wildcards.. pleae.

**Comment by xanoni** [comment]
> Why is this not in yet?

Because this is open source and nobody has sent a PR, yet, haha.

But otherwise, I agree. I've been watching this issue for years and don't even understand if this is difficult to implement. So same question.

**Comment by dannycolin** [comment]
> which is not blocked by upstream issues ("Needs: Mozilla Central").

It doesn't mean that the level of complexity is less important. When you deal with URLs, you need to validate the format of them. This includes validating the TLD of the domain which currently isn't possible without a third-party library (Public Suffix List). There's been discussions in the WECG to add a new webextension API so we can call the version already available in all browsers and avoid privacy/security issues that comes with using our own copy of the PSL.

Another pain point is dealing with URL parameters, paths and fragments. This can easily lead in a situation where the elements added by the user creates an unwanted behaviour (e.g. infinite loop between containers). Users are going to blame the extension for not handling it properly instead of them not understanding that their URL is the problem. To me, this means we can't simply add a input field with an add button and hope for the best. There need to be some UX work done in that regards before writing code. We also need test coverage for all that and contributors often ghost us when it comes the time to write tests.

In other word, this isn't a *good first bug* type of bug. We'd need someone that is comfortable with the different concepts involved. I'll be more than happy to discuss it in a 1:1 if that someone wants to contribute.

**Comment by eternal-sorrow** [comment]
>Another pain point is dealing with URL parameters, paths and fragments

Does the extension even support adding specific urls to a container? I thought it's only domains.

**Comment by eternal-sorrow** [comment]
>Support for parameters, paths and fragments can be a later improvement, or the user can rely on the workarounds mentioned above.

Again, right now there is no support for "parameters, paths and fragments" with or without workarounds. And frankly, it's not needed. And it's 100% not what this issue is about.

**Comment by dannycolin** [comment]
> Maybe a compromise can be reached in the short term, where the manual addition feature only allows adding TLDs (and subdomains).

We can't validate the TLD at the moment if it's manually added without https://github.com/w3c/webextensions/issues/231.

We could naively use https://developer.mozilla.org/en-US/docs/Web/API/URL/ for the hostname, params, etc. However, the issue I mentioned wouldn't be solved. 

@groovecoder I'm not comfortable adding this feature as something we officially support without PSL but I'll be open to accept a patch that adds an experimental section in the option page (similar to Firefox Labs) for it. What do you think about it? Would it be an acceptable compromise?

**Comment by kurahaupo** [comment]
@dannycolin I understand that this necessary to validate the _structure_ of a URL, but why does that imply needing to know which _domains_ are currently reachable? The list of valid top level domains can change dynamically, with the list comprising the "shared" public set managed by ICANN, plus domains provided by local ethernet domains (DHCP, autoconf, etc), by VPNs, by TOR, by static interface configuration, and simply by user tweaking.

Surely if a domain is not reachable, it does not matter which container is used to open it?

**Comment by tsbertalan** [comment]
Yeah, I don't see why it would be necessary to verify the content of a manually added domain. If the rule doesn't work, that's my fault; don't help me unnecessarily. I and many power users would probably prefer to be able to provide arbitrary URL regexes, but I can easily imagine that the implementation is tied in a low level that only sees the domain.
________________________________

**Comment by krokofant** [comment]
Yes please to _experimental support_, and TIL Firefox Labs! Would be great to have something to try out and evaluate rather than the user injecting custom code to workaround this.

**Comment by itsbetterhere** [comment]
Why is this a four year old enhancement?  This seems like a basic feature that should have been included at the very beginning.  All the comments as to why it hasn't been added seem to be related to underlying specifics that you don't give us access to anywaysit see.  However, it seems like you could add those details after the fact by going through the same code that you do when I pick "open this site in container x", the only extra step would be for your plugin to open the domain/url on my behalf in the first place.  Then do whatever it is you normally do under the hood.  It's just incredibly disappointing that this wasn't a requirement for the initial release.  Please just add this.

**Comment by kurahaupo** [comment]
@itsbetterhere As a fellow FMAC user, I feel the same frustration, but...

There have been some attempted patches, but none have been accepted. My understanding is that this is mostly down to the lack of available volunteer time to vet proposals, much less mentor the contributors on how to improve them.

Like any large volunteer project, Mozilla needs to be very careful about which patches they accept, since there's a strong likelihood that the person submitting it won't be sticking around to maintain it.

For me, my main obstacle is understanding the development environment. I can read code, understand it, and make changes, but I haven't a clue how to get code from github running inside my browser so I can test it.

I find myself getting immensely frustrated with "starter" guides that assume you're starting a _new_ project, rather than trying to take over an old one, and fail to give any detail of the objectives along the way.

Worse, "standard procedures" have often changed since the project started, so the guides tell you to do things that outright conflict with what's already there.

It's really demoralizing. And I'm guessing that's why there are too few volunteers.

**Comment by kurahaupo** [comment]
> For me, my main obstacle is understanding the development environment. I can read code, understand it, and make changes, but I haven't a clue how to get code from github running inside my browser so I can test it.

(I have the same problem with modifying Android apps.)

I would like an explanation that starts with:
* this is the short term this environment normally uses for an installable image (like "APK" for Android; "RPM" and "DEB" for GNU/Linux; or "JAR" for a JVM)
  * I suspect it's "plug-in" for Firefox, but there are at least 3 or 4 completely separate kinds of plug-ins for Firefox, and I don't even know what structural similarities they do or don't have: themes; search engines; 
* this is this filename suffix for a installable image
* this is how you install it somewhere where it can be tested
  * for Firefox, I suspect this is simply "copy it into X/Y/Z directory within an active profile and restart Firefox".

and then
* this is the internal structure of a installable image
  * the type of aggregator (and compressor), like TAR, ZIP, SHAR. CPIO, PAX, ISO9660, FAT32-filesystem, etc
* * details of the tree structure within it
* _these_ parts of the installable image are generated or copied from _those_ parts of the source tree.

The guide should give me enough detail  to hand-roll the plug-in image, and _not_ assume I'm going to use an "easy" tool to make it, because, frankly, that opacity is exactly where things go wrong and I get completely stymied on how to fix them.

Of course, once I understand that, I likely _will_ use your tool, but I'll need to be able to pick the installable image apart by hand, to inspect it to make sure it has everything it should have in it.

And finally
* _this_ is how you get it published so that other people can find and use it

Maybe such a guide already exists, but it's never near the top of any list of recommendations. It should have a title like "maintenance adoption guide for existing projects" or "the ultimate cheat sheet: all those details that the quick-start guides think you don't need to know".

[This guy](https://community.jenkins.io/t/first-time-user-needs-help-on-existing-project-that-already-uses-jenkins/4253) seems to have about the same experience as me, just with a different project.

**Comment by foorschtbar** [comment]
Looks like there is some progress in the PR #2710 from @dannycolin. Thank u! 🥇 

Even if #691 is older, this fix here should be also fixed with this PR.

I hope that this feature comes soon. There are countless examples where a domain-based rule alone is not enough, and we need a regex in the path part of the URL to detect right container env for this tab.

**Comment by nisc** [comment]
FoxyTab fixes this

![Image](https://github.com/user-attachments/assets/6fbe8a5e-45c3-4d54-9445-f88582d0c876)

**Comment by tsbertalan** [comment]
Not sure if it could be done privately and securely, but I could actually
imagine a gen AI enhancement to this--one big issue with allowing regexes
is that it's not clear what concrete domains would or wouldn't be covered,
esp. to a novice regex coder.

On the one hand, it might be useful to generate some positive and negative
examples so the user can immediately see the regex functioning, and
intuitively understand how it works. This might be handled by some more
simple rules-based fuzzing.

On the other hand, if the user is able to provide a few positive (and maybe
also negative) examples, AI could suggest a minimal regex that fits that
pattern.

On the third hand, it might be sufficient to allow only a tight subset of
regex, like wildcards and explicit enumerations.

**Comment by nisc** [comment]
> Not sure if it could be done privately and securely, but I could actually
> imagine a gen AI enhancement to this--one big issue with allowing regexes
> is that it's not clear what concrete domains would or wouldn't be covered,
> esp. to a novice regex coder.

I sometimes use the LLMs to put together the patterns for me, based on known domain names of a company. Probably not perfect, but better than I could do it myself in reasonable time.

**Comment by hterik** [comment]
From https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1416930533 
> In that case, you can just disable redirects in your browser temporarily. In Firefox, there's a setting called `network.http.redirection-limit` in `about:config` for that (see [here](https://kb.mozillazine.org/Network.http.redirection-limit)).
> 
> Just set it to `0` and you should be able to add the subdomain via the add-on menu.

Omg, this trick saved my day on a site that redirected before i had the chance to mark the site as "Always open in". Thanks. 👍

**Comment by jvalrog** [comment]
> From [#1670 (comment)](https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-1416930533)
> 
> > In that case, you can just disable redirects in your browser temporarily. In Firefox, there's a setting called `network.http.redirection-limit` in `about:config` for that (see [here](https://kb.mozillazine.org/Network.http.redirection-limit)).
> > Just set it to `0` and you should be able to add the subdomain via the add-on menu.
> 
> Omg, this trick saved my day on a site that redirected before i had the chance to mark the site as "Always open in". Thanks. 👍

I did this but sometimes it needs to be a combination of 0 or 1.
The thing is, it's madness.

This feature alone is totally needed to be able to fully containerize Google websites.
They do lots of redirections on many subdomains. 

Had to spend half an hour doing this and that was because I somehow found about that configuration variable on the internet.

**Comment by nisc** [comment]
> The thing is, it's madness.

I hear ya, I felt exactly the same level of frustration. But you can really work around it so well (https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-2807971883) that it no longer really affects you.

Ideally this should still be fixed, though. I agree. One extension to fix another extension is not a great solution.

**Comment by influential-eliot** [comment]
Whilst I agree with the final sentence's sentiment, @nisc, I would just bring up the use of the word "_solution_" ...
> One extension to fix another extension is not a great **solution**.
... and the overall authority of ...
> you can really work around it **so well**
... all relating to using another extension.

As you stated, it is (_at best_) a work around (_which is interminably not a **fix** - aka solution_) and one that cannot feasibly be offered by Mozilla, as it is not a work around that they '_own_' ... per se.
 
The only (_perceivable_) '_work around_' that I can see Mozilla '_owning_' is the '_temporarily change redirects_' method, which is far from user friendly.

Just putting this here to ensure that others don't take this as the (_sorry, mate_) authority that it feels like. (_sorry sorry sorry!_)

<details>
  <summary>Some notes to try to avoid irking nisc ...</summary>
  
Honestly, I am the :cursing_face::cursing_face::cursing_face::cursing_face: when I say **_any_** of this, as I cannot STAND over literal nerds online, and I hold your contribution with the greatest of respect. 🙂 👍 

But I am only _saying_ any of this to ensure that others don't just go and install another extension ... as this does come across as authoritative advice ... and I feel adding that colour should really be reserved for the Mozilla team.

I just don't believe that it is helpful to those that will perhaps find this thread when trying to sort this out to see your words and taking their authoritative nature will introduce more breakpoints and '_work_' into their general experience. The less they dislike Firefox ... the better, I think we can all agree. :)
  
</details>

Equally, speaking personally, I would not be willing to introduce said browser extension into my work setup, which is my primary use case for MAC. As they're the most productive tool I have as a consultant working with many clients. So keeping my browser experience as threadbare as I can (_with regard to extensions, scripts, and suchlike_) is vital to keeping things ticking along. :)

**Comment by mm12** [comment]
Being able to, at the very least, toggle between strictly a domain (say, `www.example.com`) and "any subdomain" (say, `ANY.example.com`) would be  HUGE help. While more advanced rules would be nice to have, I see subdomain rules are the bare minimum. This is because many sites give specific users/creators a subdomain for their site. For example, `user.bsky.social` - in this case it redirects, but the idea is the same.

**Comment by influential-eliot** [comment]
I believe that it already does distinguish between subdomains, does it not, @mm12 ?

> Being able to, at the very least, toggle between strictly a domain (say, `www.example.com`) and "any subdomain" (say, `ANY.example.com`) would be HUGE help. While more advanced rules would be nice to have, I see subdomain rules are the bare minimum. This is because many sites give specific users/creators a subdomain for their site. For example, `user.bsky.social` - in this case it redirects, but the idea is the same.

Maybe I've been doing it wrong, though.

-----

Oh, and, if anyone is interested in one way I've used this lack of functionality to turn it into a feature ....
<details>
  <summary> ... read on ...</summary>

Basically, I needed to have fast ways to open common domains in different containers based on the client (_literally a client, or just personal / work / etc_) and user aligned with it.

So, I started to make fake .local domains, and use the redirector addon to then redirect those very specific domains back out to their intended when they're in their correct container.

That way I can add bookmarks and suchlike and know that they'll always open in a given container. 

It's actually fully mapped out on the userscript page ( _[bustoutmoz](https://greasyfork.org/en/scripts/543865)_ ) that I made to minimally to stop Mozilla (_in a rather predatorial fashion, given the main purpose of MAC_) breaking the MAC sanctity by opening moz.org links in whichever container it feels like it. Dropping cookies, and having access to which site a user previously came from, etc ...

... anyway, not got time to go more in to it now, but if there's interest for me to put more effort into something around this, I will. :)
  
</details>

**Comment by mm12** [comment]
> I believe that it already does distinguish between subdomains, does it not, @mm12 

@influential-eliot  yes, it does, and I want it to _not_ do that. I would like `123.example.com` and `456.example.com` to be treated the same.

**Comment by influential-eliot** [comment]
Aha, gotcha ... totally get it ... pain in the neck on the very odd occasion indeed. 👍 

> [@influential-eliot](https://github.com/influential-eliot) yes, it does, and I want it to _not_ do that. I would like `123.example.com` and `456.example.com` to be treated the same.

**Comment by VL4DST3R** [comment]
How is this still an issue 5 years later? Is this extension dead/abandoned by mozilla?

**Comment by nisc** [comment]
> How is this still an issue 5 years later? Is this extension dead/abandoned by mozilla?

We're the power user minority I guess.

**Comment by VL4DST3R** [comment]
> We're the power user minority I guess.

Given that most "big" site logins like google/youtube/microsoft - not to mention virtually all online stores - do the multiple redirects thing, I can't imagine how it even left their internal testing unaddressed.

Even if it is some obscure technical reason (read:bull) why it cannot be done, at the very least *some* attempt at mitigating it would have been nice. Basically any one of the suggestions mentioned above would have gone a long way from making this an annoying game of whack-a-mole or always having to manually toggle back and forth the limit to designated sites option.

**Comment by kurahaupo** [comment]
@VL4DST3R nice of you to volunteer to (pay someone to) write a patch.

Seriously, this extension fulfils its intended purpose: it's a worked example that demonstrates how to use Firefox's Container API.

Unfortunately nobody else has written a “better” plug-in, which leaves frustrated ordinary users trying to use this one. You're welcome to fork it and invite patches, or even apply some of the ones that have been suggested in other tickets here.

**Comment by VL4DST3R** [comment]
> nice of you to volunteer to (pay someone to) write a patch.

Last I checked this is made by **mozilla**, not a single dev doing it out of passion with no financial backing.

As long as Mozilla is receiving funding and gathering plenty of *our* user data, me asking that at least their software be feature complete - and more importantly adapted to function *properly* in today's internet landscape - is *not* an overstepping of some... imaginary boundary you seem to have set up for them.

**Comment by kurahaupo** [comment]
> Last I checked this is made by **mozilla**, not a single dev doing it out of passion with no financial backing.

Let's be clear, "this" is a plug-in, not a core part of Firefox.

And when I checked…

Looking at the email addresses of the 1228 commits in the `main` branch, there have been 194 commits made by identifiable Mozilla staff, 71 by GitHub users with masked email addresses, and 963 made by others.

Sure, Mozilla started this project; other people have extended it to make it what it is now. Mozilla is no longer even the majority contributor.

Yes, most of this project is indeed made by individuals doing it out of passion with no financial backing.

(Yes I'm sure some of those commits are by Mozilla staff using private accounts, but a majority seems unlikely.)

> As long as Mozilla is _[…]_
> gathering plenty of _our_ user data

So you're accusing Mozilla of lying about their Privacy Principles?
https://www.mozilla.org/en-US/privacy/principles/

If you really believe that, why are you using any Mozilla products?

And call your lawyer and ask them about libel.

> receiving funding
_[…]_
> _not_ an overstepping of some... imaginary boundary you seem to have set up for them.

Regardless of payments, it's an open source project. Like every open source project, you're welcome to fork it if you don't like the direction the project is taking.

It's not a question of a boundary (imaginary or otherwise) _for Mozilla_; rather it's my expectation for _every_ open source project, and of the users of those projects.

I take issue with the "I'm entitled to something for nothing" attitude often exhibited by many users of open source software. Users could at least show some gratitude; "why haven't you fixed it" is the opposite of gratitude.

----

Lastly, Mozilla's funding has fallen in recent years, and as a result they have reduced their paid staff by about 90% since 2020. There are currently exactly zero paid staff even monitoring at this project, much less actively contributing. We have just one overworked unpaid volunteer ex-staffer who has contact details to ask paid staff to approve pull requests, and maybe half a dozen folk like me who are trying to write patches.

Given that the main bottleneck is actually getting pull requests approved, simply forking the project is indeed the most effective approach. If folk are willing to be helpful rather than belligerent, then maybe I'll get to doing that myself. Or if someone else wants to do it then I'll support them.

**Comment by VL4DST3R** [comment]
My issue was specifically pointed at *Mozilla* throughout all of my replies, who advertised a feature and then abandoned it. Not at the volunteers that contributed to the project. Sorry if I sounded like I was implying otherwise, it was absolutely not my intention. I have nothing but respect for people who can and do voluntarily contribute to projects like this.

> There are currently exactly zero paid staff even monitoring at this project, much less actively contributing.

This is valuable info. Leading with this to my original question if the project is "dead/abandoned by mozilla" would have saved everyone the time and energy. Sad if this is indeed the case but at least we know what to expect.

> So you're accusing Mozilla of lying about their Privacy Principles?

Not exactly, AFAIK there's nothing conclusive to put them in the spotlight _yet_ but the controversy from some months ago ([read here](https://arstechnica.com/tech-policy/2025/02/firefox-deletes-promise-to-never-sell-personal-data-asks-users-not-to-panic/), [or here](https://techcrunch.com/2025/02/28/mozilla-responds-to-backlash-over-new-terms-saying-its-not-using-peoples-data-for-ai/)), combined with the fact that they *do* in fact collect/forward telemetry data from you by default with things like Pocket (or *did* since its shutting down), search suggestions, sponsored shortcuts, etc. still leaves to be desired.

> If you really believe that, why are you using any Mozilla products?

I'm not actually using Firefox directly anymore, I'm maining Waterfox at the moment which more closely lines up with what I was personally expecting from Firefox's privacy policy to begin with.

**Comment by influential-eliot** [comment]
Might be worth moving a lot of this to a 'discussion' ... no?

**Comment by grahamperrin** [comment]
> moving

+100

**Comment by influential-eliot** [comment]
Just popping in to state that there's a very valid reason this should ramp up a little, and that is the prevalence of Web Apps.
 
If you use a Web App, and wish to have MAC enabled for multiple '_tenancies_' (_or whatever_), then only the valid/whitelisted URLs/domains, will open in that PWA. So, for a non-advanced user (_who is unlikely to engage in some of the workarounds in this thread_) it's just not possible for them to (_for example_) add their second Outlook tenancy, even via a fake local domain.

**Comment by E14** [comment]
> Just popping in to state that there's a very valid reason this should ramp up a little, and that is the prevalence of Web Apps.
> 
> If you use a Web App, and wish to have MAC enabled for multiple '_tenancies_' (_or whatever_), then only the valid/whitelisted URLs/domains, will open in that PWA. So, for a non-advanced user (_who is unlikely to engage in some of the workarounds in this thread_) it's just not possible for them to (_for example_) add their second Outlook tenancy, even via a fake local domain.

There seems to be an ongoing effort to get more support for this plugin, see also #2814, I just hope it doesn't go the way of  tab-groups, but apparently this time we have at least some invested supporters internally to Mozilla.

**Comment by influential-eliot** [comment]
Well ... tab groups did come back in the new flavour, innit.

> > Just popping in to state that there's a very valid reason this should ramp up a little, and that is the prevalence of Web Apps.
> > If you use a Web App, and wish to have MAC enabled for multiple '_tenancies_' (_or whatever_), then only the valid/whitelisted URLs/domains, will open in that PWA. So, for a non-advanced user (_who is unlikely to engage in some of the workarounds in this thread_) it's just not possible for them to (_for example_) add their second Outlook tenancy, even via a fake local domain.
> 
> There seems to be an ongoing effort to get more support for this plugin, see also [#2814](https://github.com/mozilla/multi-account-containers/discussions/2814), I just hope it doesn't go the way of tab-groups, but apparently this time we have at least some invested supporters internally to Mozilla.


However ... that old interface was AMAZING, am I right?

**Comment by jacobckelly** [comment]
After having to hit stop on multiple domains that a Microsoft login hits to get that Container working, I'm going to add my voice for this. Being able to manually add and edit domains is a key feature.

**Comment by LordGyarados** [comment]
> After having to hit stop on multiple domains that a Microsoft login hits to get that Container working, I'm going to add my voice for this. Being able to manually add and edit domains is a key feature.

Same. Having to add a separate entry for every bloody redirect a site throws at us is tiring, and in some cases, plain impossible. By the time you try to add the current redirecting page to a container through the extension's GUI, it has already redirected to another page.

EDIT: Note that there **is** a workaround for this, although it's slightly silly: take the URL of the redirecting page, strip it of all its parameters so you are left with only the domain + page, and then visit the resulting URL. You will likely land on a 404 page as a result of a php error. You now have the time to add this page to a container.

**Comment by foorschtbar** [comment]
> > After having to hit stop on multiple domains that a Microsoft login hits to get that Container working, I'm going to add my voice for this. Being able to manually add and edit domains is a key feature.
> 
> Same. Having to add a separate entry for every bloody redirect a site throws at us is tiring, and in some cases, plain impossible. By the time you try to add the current redirecting page to a container through the extension's GUI, it has already redirected to another page.

Yes and the meantime use Containerise. With the regex based rules i have a perfect working setup for multiple microsoft tenants running - when they have a tenant marker in the url.

**Comment by eternal-sorrow** [comment]
>Yes and the meantime use Containerise

It doesn't work well with Temporary Containers.

**Comment by influential-eliot** [comment]
Have a question ... purely out of curiousity ... [so I've started a discussion on this in discussions](https://github.com/mozilla/multi-account-containers/discussions/2857), @jacobckelly 🙂 ... just to stop me derailing this.

> After having to hit stop on multiple domains that a Microsoft login hits to get that Container working, I'm going to add my voice for this. Being able to manually add and edit domains is a key feature.

**Comment by kurahaupo** [comment]
@LordGyarados 

> Same. Having to add a separate entry for every bloody redirect a site throws at us is tiring, and in some cases, plain impossible. By the time you try to add the current redirecting page to a container through the extension's GUI, it has already redirected to another page.

You can prevent that by setting `network.http.redirection-limit` in `about:config` to `0`.
That will give you the first domain.
Then set it to `1` and re-load to get to the next domain.
Then re-load after each redirection to iterate the domains, recording as you go. 

(PS: I agree that it's bug and this hack shouldn't be needed.)
(PPS: don't forget to set `network.http.redirection-limit` back to its default when you're done.)

**Comment by RemiCardona** [comment]
if there was a way to pin comment https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-3942409025 to the top, 'cause that's the only "proper" solution ATM.

**Comment by SoCuul** [comment]
I'm not sure why they haven't implemented a proper way to do this, but here's my try at a workaround:

1. Set a default container for a site, for later reference to get the container id
2. Open `about:debugging`, scroll down to "Firefox Multi-Account Containers" and click **Inspect**
3. Click on the storage tab, then on the left expand "Extension Storage" and click on the `moz-extension://` part
4. Find the container entry for the site that you associated earlier, and grab the container id from it.
5. Go back to the console and run the following javascript (after writing `allow pasting` if needed)

```js
window.assignManager._setOrRemoveAssignment(0, "https://mysite.com", "containerId")
```

Also made a quick video demonstration for anyone who would prefer it:

https://github.com/user-attachments/assets/85f1b6bd-3e4f-47a4-af78-394eb8e88ffb

---

I would make a pull request to implement this functionality, but looking at the status of this repo I feel like it'd be a waste of time considering it wouldn't get merged, or even looked at.

**Comment by sshquack** [comment]
Dear Mozilla devs, Pretty please add this feature to add a specific site (ideally using a pattern). Currently [network.http.redirection-limit 0](https://github.com/mozilla/multi-account-containers/issues/1670#issuecomment-3942409025) is the only fix.

Here is all similar related issues:
- https://github.com/mozilla/multi-account-containers/issues/691
- https://github.com/mozilla/multi-account-containers/issues/473
- https://github.com/mozilla/multi-account-containers/issues/473
- and many more

Thank you.

**Comment by wrouesnel** [comment]
Bump for this feature: I was blown away this did not exist.

**Comment by MrBrax** [comment]
indeed, this addon seems to be quite abandoned, shame for such basic features

**Comment by gavinbeatty** [comment]
> window.assignManager._setOrRemoveAssignment(0, "https://mysite.com", "containerId")

Thanks for this, however when I tried following all the instructions just now, I received the following error and no change occurred.

> Error: QuotaExceededError: storage.sync API call exceeded its quota limitations

**Comment by kurahaupo** [comment]
@gavinbeatty see [adjusting your storage quota](https://support.mozilla.org/tn/questions/1353337)
